src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateManager.cs

```diff
@@ -308,6 +308,18 @@ private void VerifyProjectDiagnosticStates(IEnumerable<StateSet> stateSets)
 
                 VerifyUniqueStateNames(hostStates.Concat(stateSets));
             }
+
+            private readonly struct HostAnalyzerStateSetKey
```

Comment:
❗ This needs to implement `IEquatable<HostAnalyzerStateSetKey>` or it will fall back to the very slow reflection-based implementation from ValueType.